include/graaflib/algorithm/clique/bron_kerbosch.tpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <graaflib/algorithm/clique/bron_kerbosch.h>
+
+#include <iostream>
+#include <iterator>
+#include <optional>
+#include <ranges>
+#include <unordered_set>
+#include <vector>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+void do_get_intersection(std::unordered_set<vertex_id_t>& s1,
+                         std::unordered_set<vertex_id_t>& s2,
+                         std::unordered_set<vertex_id_t>& intersection) {
+  for (const auto& vertex : s1) {
+    if (s2.find(vertex) != s2.end()) {
+      intersection.insert(vertex);
+    }
+  }
+}
+
+template <typename V, typename E>
+void do_bron_kerbosch_maximal_clique(
+    std::vector<vertex_id_t>& clique, std::unordered_set<vertex_id_t>& vertices,
+    std::unordered_set<vertex_id_t>& excluded_vertices,
+    std::vector<std::vector<vertex_id_t>>& maximal_cliques,
+    const graph<V, E, graph_type::UNDIRECTED>& graph) {
+  if (vertices.empty() && excluded_vertices.empty()) {
+    maximal_cliques.push_back(clique);
+    return;
+  }
+
+  // union_set = candidate_vertices ⋃ excluded_vertices
+  auto union_set = vertices;
+  for (const auto& vertex : excluded_vertices) union_set.insert(vertex);
+
+  // Get pivot vertex with most amount of neighbors
+  // to reduce recursion calls
+  auto pivot_vertex = *union_set.begin();
+  for (const auto& vertex : union_set) {
+    if (graph.get_neighbors(pivot_vertex).size() <
+        graph.get_neighbors(vertex).size()) {
+      pivot_vertex = vertex;
+    }
+  }
```

Comment:
This works and should be fine to leave as-is. However, if you are up for it this sounds like a great use-case to use C++ 20's [std::ranges::max_element](https://en.cppreference.com/w/cpp/algorithm/ranges/max_element) in combination with a projection.

```c++
const auto pivot_vertex = *std::ranges::max_element(union_set, {}, [&graph](vertex_id_t id) {
  return graph.get_neighbors(id).size();
};
```